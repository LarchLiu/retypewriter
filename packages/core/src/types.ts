export interface InsertPatch {
  type: 'insert'
  from: number
  text: string
}

export interface RemovalPatch {
  type: 'removal'
  from: number
  length: number
}

export type Patch = InsertPatch | RemovalPatch

export interface Snapshot {
  content: string
  options?: SnapshotOptions
}

export interface SnapshotOptions {
  wait?: number
}

export interface AnimatorStepInsert {
  type: 'insert' | 'delete'
  cursor: number
  content: string
  char: string
}

export interface AnimatorStepRemoval {
  type: 'removal'
  cursor: number
  content: string
}

export interface AnimatorStepInit {
  type: 'init'
  content: string
}

export interface AnimatorStepPatch {
  type: 'new-patch'
  patch: Patch
  index: number
}

export interface AnimatorStepSnap {
  type: 'new-snap'
  snap: Snapshot
  index: number
}

export type AnimatorStep = AnimatorStepInsert | AnimatorStepRemoval | AnimatorStepInit | AnimatorStepPatch | AnimatorStepSnap
