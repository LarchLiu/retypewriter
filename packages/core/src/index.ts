export * from './steps'
export * from './types'
export * from './patch'
export * from './snaps'
export * from './slicing'
export * from './timing'
export * from './typewriter'
