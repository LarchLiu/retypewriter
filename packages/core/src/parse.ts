import YAML from 'js-yaml'
import type { ParsedHead, ParsedSnaphot, Snapshot, SnapshotOptions } from './types'

export const SNAP_EXT = '.retypewriter'
export const SNAP_HEADING = 'reTypewriter Snapshots v1\n'
export const SNAP_SEPERATOR_PRE = '-'.repeat(2)
export const SNAP_SEPERATOR_POST = '-'.repeat(10)
export const SNAP_SEPERATOR = `${SNAP_SEPERATOR_PRE}--${SNAP_SEPERATOR_POST}`
export const SNAP_SEPERATOR_OPTIONS = '-----options--'
export const SNAP_SEPERATOR_MATCHER = new RegExp(`\\n?${SNAP_SEPERATOR_PRE}[\\w-]{2}${SNAP_SEPERATOR_POST}\\n`, 'g')
export const SNAP_SEPERATOR_OPTIONS_MATCHER = new RegExp(`\\n?${SNAP_SEPERATOR_OPTIONS}\\n`, 'g')

export function parseOptions(raw: string): SnapshotOptions | undefined {
  raw = raw.trim()
  if (!raw)
    return undefined
  return (raw.startsWith('{'))
    ? JSON.parse(raw)
    : YAML.load(raw)
}

export function stringifySnapshots(snapshots: Snapshot[], useYaml = true) {
  return [
    SNAP_HEADING,
    ...snapshots
      .flatMap((snap, i) => [
        SNAP_SEPERATOR_PRE + (i + 1).toString().padStart(2, '0') + SNAP_SEPERATOR_POST,
        snap.content,
        ...(
          snap.options
            ? [
              SNAP_SEPERATOR_OPTIONS,
              useYaml
                ? YAML.dump(snap.options, { indent: 2 }).trimEnd()
                : Object.keys(snap.options).length > 1
                  ? JSON.stringify(snap.options, null, 2)
                  : JSON.stringify(snap.options),
            ]
            : []
        ),
      ]),
    SNAP_SEPERATOR,
    '',
  ].join('\n')
}

export function parseSnapshots(raw: string) {
  if (!raw.startsWith(SNAP_HEADING))
    throw new SyntaxError('Invalid snapshot file')
  if (!raw.endsWith('\n'))
    raw += '\n'

  const head: ParsedHead = {}
  const snaps: ParsedSnaphot[] = []

  SNAP_SEPERATOR_MATCHER.lastIndex = 0
  const matches = Array.from(raw.matchAll(SNAP_SEPERATOR_MATCHER))

  matches.forEach((match, idx) => {
    const start = match.index!

    if (idx === 0) {
      const content = raw.slice(0, start)
      head.options = parseOptions(content.slice(SNAP_HEADING.length))
    }

    const next = matches[idx + 1]
    const end = next?.index ?? raw.length
    const bodyStart = start + match[0].length
    const withOptions = raw.slice(bodyStart, end)
    const parts = withOptions.split(SNAP_SEPERATOR_OPTIONS_MATCHER)
    const body = parts[0]
    const bodyEnd = bodyStart + body.length
    const snap: ParsedSnaphot = {
      raw: raw.slice(start, end),
      start,
      end,
      body,
      bodyEnd,
      bodyStart,
    }
    const optionsRaw = parts[1]?.trim()
    if (optionsRaw) {
      snap.optionsRaw = optionsRaw
      snap.options = parseOptions(optionsRaw)
    }
    snaps.push(snap)
  })

  // remove tailing separator
  snaps.pop()

  return {
    head,
    snapshots: snaps,
  }
}
