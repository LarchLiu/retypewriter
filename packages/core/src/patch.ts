import type { Diff } from 'diff-match-patch'
import { diff_match_patch as DMP } from 'diff-match-patch'
import type { Patch } from './types'

export function diff(a: string, b: string): Diff[] {
  const differ = new DMP()
  const delta = differ.diff_main(a, b)
  differ.diff_cleanupSemantic(delta)
  return delta
}

export function calculatePatch(diff: Diff[]): Patch[] {
  const patches: Patch[] = []

  let cursor = 0
  for (const change of diff) {
    if (change[0] === 0) {
      cursor += change[1].length
    }
    else if (change[0] === -1) {
      const length = change[1].length
      patches.push({
        type: 'removal',
        cursor: cursor + length,
        length,
      })
    }
    else if (change[0] === 1) {
      const content = change[1]
      const tailingNewLines = content.match(/\n*$/m)?.[0] || ''
      const rest = tailingNewLines.length
        ? content.slice(0, -tailingNewLines.length)
        : content
      // if there is new lines in the patch, we should type them first
      if (tailingNewLines.length) {
        patches.push({
          type: 'insert',
          cursor,
          content: tailingNewLines,
        })
      }
      if (rest.length) {
        patches.push({
          type: 'insert',
          cursor,
          content: rest,
        })
      }
      cursor += content.length
    }
    else {
      throw new Error('unknown change type')
    }
  }
  return patches
}
