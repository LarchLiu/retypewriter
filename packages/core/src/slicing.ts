import type { Slice } from './types'

const pairs = [
  ['{', '}'],
  ['(', ')'],
  ['[', ']'],
  ['<', '>'],
  ['`', '`'],
  ['\'', '\''],
  ['"', '"'],
]

const chars = '(){}\\[\\]<>\'"`'
const splitRegex = new RegExp(`(\\s*[${chars}]\\s*)`)

export function sliceInput(input: string): Slice[] {
  if (!input)
    return []
  const tailingNewLines = input.match(/\n+$/m)?.[0] || ''
  const rest = tailingNewLines.length
    ? input.slice(0, -tailingNewLines.length)
    : input

  const strings = rest.split(splitRegex).filter(Boolean)

  let index = 0
  const slices = strings
    .map((s, idx) => {
      const trimmed = s.trim()
      const pair = pairs.find(i => i[1] === trimmed)
      const order = (pair && strings[idx - 2]?.trim() === pair[0])
        ? idx - 2 // paired with previous
        : idx

      const _index = index
      index += s.length

      return {
        content: s,
        order,
        cursor: _index,
      }
    })
    .sort((a, b) => a.order - b.order)

  let cursor = 0
  slices.forEach((i) => {
    i.cursor = Math.min(i.cursor, cursor)
    cursor += i.content.length
  })

  if (tailingNewLines.length) {
    slices.unshift({
      content: tailingNewLines,
      order: -1,
      cursor: 0,
    })
  }

  return slices
}

export function applySlice(slices: Slice[]) {
  const result = slices.reduce((acc, i) => {
    const { content: contet, cursor } = i
    return acc.slice(0, cursor) + contet + acc.slice(cursor)
  }, '')
  return result
}
