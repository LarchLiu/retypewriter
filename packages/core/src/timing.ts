export const timeRanges: [string, number, number][] = [
  // default
  ['', 0, 100],
  // commons
  [' ', 0, 40],
  ['adefijkloosu', 0, 50],
  ['bcghmnpqrtvwxyz', 5, 70],
  // uppercase
  ['adefijkloosu'.toUpperCase(), 0, 80],
  ['bcghmnpqrtvwxyz'.toUpperCase(), 5, 100],
  // a bit far
  ['[];\',./-=', 10, 120],
  // need shift
  ['<>?:"{}|\\_+()*', 20, 140],
  // a bit far and shift
  ['!~`#$%^&', 20, 180],
]

export function randRange(min: number, max: number): number {
  return Math.random() * (max - min) + min
}

export function getTimeout(char: string, multiplier = 1): number {
  const [, min, max] = timeRanges.find(range => range[0].includes(char)) || timeRanges[0]
  return randRange(min, max) * multiplier
}
