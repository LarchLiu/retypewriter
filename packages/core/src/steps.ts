import { calculatePatch, diff } from './patch'
import { sliceInput } from './slicing'
import type { AnimatorStep, Patch, Snapshot } from './types'

export function *patchSteps(input: string, patches: Patch[]): Generator<AnimatorStep> {
  let output = input
  let cursor = 0

  for (let index = 0; index < patches.length; index++) {
    const patch = patches[index]

    yield {
      type: 'patch-start',
      patch,
      index,
      total: patches.length,
    }

    if (patch.type === 'insert') {
      cursor = patch.cursor
      const head = output.slice(0, patch.cursor)
      const tail = output.slice(patch.cursor)
      for (const { char, output, cursor: delta } of animateInsertionSlices(patch.content)) {
        yield {
          type: 'insert',
          char,
          cursor: cursor + delta,
          content: head + output + tail,
        }
      }
      output = head + patch.content + tail
    }
    else if (patch.type === 'removal') {
      cursor = patch.cursor - patch.length
      const head = output.slice(0, cursor)
      const tail = output.slice(patch.cursor)
      const selection = output.slice(cursor, patch.cursor)
      for (let i = selection.length - 1; i >= 0; i--) {
        yield {
          type: 'removal',
          cursor: cursor + i,
          content: head + selection.slice(0, i) + tail,
        }
      }
      output = head + tail
    }

    yield {
      type: 'patch-finish',
      content: output,
      index,
      total: patches.length,
    }
  }
}

export function *animateSteps(snapshots: Snapshot[]): Generator<AnimatorStep> {
  let lastContent: string | undefined
  const copy = [...snapshots]
  for (let index = 0; index < copy.length; index++) {
    const snap = copy[index]
    if (lastContent == null) {
      lastContent = snap.content
      yield {
        type: 'init',
        content: lastContent,
      }
      continue
    }

    yield {
      type: 'snap-start',
      snap,
      index,
      total: copy.length,
    }

    const steps = stepsTo(lastContent, snap.content)
    for (const step of steps)
      yield step

    yield {
      type: 'snap-finish',
      content: snap.content,
      index,
      total: copy.length,
    }

    lastContent = snap.content
  }
}

export function *animateInsertionSlices(input: string) {
  const slices = sliceInput(input)
  let output = ''
  for (const { content, cursor } of slices) {
    const head = output.slice(0, cursor)
    const tail = output.slice(cursor)

    let body = ''
    for (const char of content) {
      body += char
      yield {
        char,
        output: head + body + tail,
        cursor: cursor + body.length,
      }
    }
    output = head + content + tail
  }
}

export function stepsTo(input: string, output: string) {
  const delta = diff(input, output)
  const patches = calculatePatch(delta)
  return patchSteps(input, patches)
}

export function applyPatches(input: string, patches: Patch[]) {
  let output = input
  for (const patch of patchSteps(input, patches)) {
    if (patch.type === 'patch-finish')
      output = patch.content
  }
  return output
}
