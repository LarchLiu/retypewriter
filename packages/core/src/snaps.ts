import type { AnimatorStep, Snapshot } from './types'
import { stepsTo } from './steps'
import type { TypewriterOptions } from './typewriter'
import { typingAnimator } from './typewriter'
import { SNAP_EXT, parseSnapshots, stringifySnapshots } from './parse'

export class Snapshots extends Array<Snapshot> {
  constructor(...args: Snapshot[]) {
    super(...args)
  }

  get last() {
    return this[this.length - 1]
  }

  get first() {
    return this[0]
  }

  move(from: number, to: number) {
    if (from === to)
      return

    const element = this[from]
    this.splice(from, 1)
    this.splice(to, 0, element)
  }

  toString(useYaml = true) {
    return stringifySnapshots(this, useYaml)
  }

  fromString(raw: string) {
    const { snapshots: parsed } = parseSnapshots(raw)
    this.length = 0
    parsed.forEach((p) => {
      this.push({
        content: p.body,
        options: p.options,
      })
    })
    return this
  }

  static fromString(raw: string) {
    return new Snapshots().fromString(raw)
  }

  *steps(): Generator<AnimatorStep> {
    let lastContent: string | undefined
    const copy = [...this]
    for (let index = 0; index < copy.length; index++) {
      const snap = copy[index]
      if (lastContent == null) {
        lastContent = snap.content
        yield {
          type: 'init',
          content: lastContent,
        }
        continue
      }

      yield {
        type: 'new-snap',
        snap,
        index,
      }
      const steps = stepsTo(lastContent, snap.content)
      for (const step of steps)
        yield step

      lastContent = snap.content
    }
  }

  typewriter(options?: TypewriterOptions) {
    return typingAnimator(this.steps(), options)
  }
}

export type SnapshotFallbackLoader = (id: string) => Snapshots | undefined | Promise<Snapshots | undefined>
export interface SnapshotManagerOptions {
  ensureFallback?: SnapshotFallbackLoader
}

export class SnapshotManager extends Map<string, Snapshots> {
  constructor(
    public options: SnapshotManagerOptions = {},
  ) {
    super()
  }

  async ensure(
    id: string,
    load = this.options.ensureFallback,
  ) {
    if (!this.has(id))
      this.set(id, await load?.(id) || new Snapshots())
    return this.get(id)!
  }
}

export function getSnapshotPath(id: string) {
  if (id.endsWith(SNAP_EXT))
    return id
  return id + SNAP_EXT
}

export function getOriginalFilePath(id: string) {
  if (id.endsWith(SNAP_EXT))
    return id.slice(0, -SNAP_EXT.length)
  return undefined
}
