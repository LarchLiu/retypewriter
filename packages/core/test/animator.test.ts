import { expect, it } from 'vitest'
import { simpleAnimator } from '../src'
import { input, output } from './fixture'

it('animator', () => {
  const animator = simpleAnimator(input, output)

  expect([...animator]).toMatchSnapshot('animator')
})
