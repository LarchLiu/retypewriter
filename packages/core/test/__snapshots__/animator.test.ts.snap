// Vitest Snapshot v1

exports[`animator > animator 1`] = `
[
  {
    "index": 0,
    "patch": {
      "from": 75,
      "length": 29,
      "type": "removal",
    },
    "type": "new-patch",
  },
  {
    "content": "
import { describe, expect, it } from 'vitest'
import { one } from '../src

describe('should', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
    "cursor": 74,
    "type": "removal",
  },
  {
    "content": "
import { describe, expect, it } from 'vitest'
import { one } from '../sr

describe('should', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
    "cursor": 73,
    "type": "removal",
  },
  {
    "content": "
import { describe, expect, it } from 'vitest'
import { one } from '../s

describe('should', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
    "cursor": 72,
    "type": "removal",
  },
  {
    "content": "
import { describe, expect, it } from 'vitest'
import { one } from '../

describe('should', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
    "cursor": 71,
    "type": "removal",
  },
  {
    "content": "
import { describe, expect, it } from 'vitest'
import { one } from '..

describe('should', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
    "cursor": 70,
    "type": "removal",
  },
  {
    "content": "
import { describe, expect, it } from 'vitest'
import { one } from '.

describe('should', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
    "cursor": 69,
    "type": "removal",
  },
  {
    "content": "
import { describe, expect, it } from 'vitest'
import { one } from '

describe('should', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
    "cursor": 68,
    "type": "removal",
  },
  {
    "content": "
import { describe, expect, it } from 'vitest'
import { one } from 

describe('should', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
    "cursor": 67,
    "type": "removal",
  },
  {
    "content": "
import { describe, expect, it } from 'vitest'
import { one } from

describe('should', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
    "cursor": 66,
    "type": "removal",
  },
  {
    "content": "
import { describe, expect, it } from 'vitest'
import { one } fro

describe('should', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
    "cursor": 65,
    "type": "removal",
  },
  {
    "content": "
import { describe, expect, it } from 'vitest'
import { one } fr

describe('should', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
    "cursor": 64,
    "type": "removal",
  },
  {
    "content": "
import { describe, expect, it } from 'vitest'
import { one } f

describe('should', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
    "cursor": 63,
    "type": "removal",
  },
  {
    "content": "
import { describe, expect, it } from 'vitest'
import { one } 

describe('should', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
    "cursor": 62,
    "type": "removal",
  },
  {
    "content": "
import { describe, expect, it } from 'vitest'
import { one }

describe('should', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
    "cursor": 61,
    "type": "removal",
  },
  {
    "content": "
import { describe, expect, it } from 'vitest'
import { one 

describe('should', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
    "cursor": 60,
    "type": "removal",
  },
  {
    "content": "
import { describe, expect, it } from 'vitest'
import { one

describe('should', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
    "cursor": 59,
    "type": "removal",
  },
  {
    "content": "
import { describe, expect, it } from 'vitest'
import { on

describe('should', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
    "cursor": 58,
    "type": "removal",
  },
  {
    "content": "
import { describe, expect, it } from 'vitest'
import { o

describe('should', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
    "cursor": 57,
    "type": "removal",
  },
  {
    "content": "
import { describe, expect, it } from 'vitest'
import { 

describe('should', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
    "cursor": 56,
    "type": "removal",
  },
  {
    "content": "
import { describe, expect, it } from 'vitest'
import {

describe('should', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
    "cursor": 55,
    "type": "removal",
  },
  {
    "content": "
import { describe, expect, it } from 'vitest'
import 

describe('should', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
    "cursor": 54,
    "type": "removal",
  },
  {
    "content": "
import { describe, expect, it } from 'vitest'
import

describe('should', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
    "cursor": 53,
    "type": "removal",
  },
  {
    "content": "
import { describe, expect, it } from 'vitest'
impor

describe('should', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
    "cursor": 52,
    "type": "removal",
  },
  {
    "content": "
import { describe, expect, it } from 'vitest'
impo

describe('should', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
    "cursor": 51,
    "type": "removal",
  },
  {
    "content": "
import { describe, expect, it } from 'vitest'
imp

describe('should', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
    "cursor": 50,
    "type": "removal",
  },
  {
    "content": "
import { describe, expect, it } from 'vitest'
im

describe('should', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
    "cursor": 49,
    "type": "removal",
  },
  {
    "content": "
import { describe, expect, it } from 'vitest'
i

describe('should', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
    "cursor": 48,
    "type": "removal",
  },
  {
    "content": "
import { describe, expect, it } from 'vitest'


describe('should', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
    "cursor": 47,
    "type": "removal",
  },
  {
    "content": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
    "cursor": 46,
    "type": "removal",
  },
  {
    "index": 1,
    "patch": {
      "from": 89,
      "length": 8,
      "type": "removal",
    },
    "type": "new-patch",
  },
  {
    "content": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('exporte', () => {
    expect(one).toEqual(1)
  })
})
",
    "cursor": 88,
    "type": "removal",
  },
  {
    "content": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('export', () => {
    expect(one).toEqual(1)
  })
})
",
    "cursor": 87,
    "type": "removal",
  },
  {
    "content": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('expor', () => {
    expect(one).toEqual(1)
  })
})
",
    "cursor": 86,
    "type": "removal",
  },
  {
    "content": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('expo', () => {
    expect(one).toEqual(1)
  })
})
",
    "cursor": 85,
    "type": "removal",
  },
  {
    "content": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('exp', () => {
    expect(one).toEqual(1)
  })
})
",
    "cursor": 84,
    "type": "removal",
  },
  {
    "content": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('ex', () => {
    expect(one).toEqual(1)
  })
})
",
    "cursor": 83,
    "type": "removal",
  },
  {
    "content": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('e', () => {
    expect(one).toEqual(1)
  })
})
",
    "cursor": 82,
    "type": "removal",
  },
  {
    "content": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('', () => {
    expect(one).toEqual(1)
  })
})
",
    "cursor": 81,
    "type": "removal",
  },
  {
    "index": 2,
    "patch": {
      "from": 81,
      "text": "one",
      "type": "insert",
    },
    "type": "new-patch",
  },
  {
    "char": "o",
    "content": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('o', () => {
    expect(one).toEqual(1)
  })
})
",
    "cursor": 82,
    "type": "insert",
  },
  {
    "char": "n",
    "content": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('on', () => {
    expect(one).toEqual(1)
  })
})
",
    "cursor": 83,
    "type": "insert",
  },
  {
    "char": "e",
    "content": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('one', () => {
    expect(one).toEqual(1)
  })
})
",
    "cursor": 84,
    "type": "insert",
  },
  {
    "index": 3,
    "patch": {
      "from": 122,
      "text": "    expect(two).toEqual(2)
",
      "type": "insert",
    },
    "type": "new-patch",
  },
  {
    "char": " ",
    "content": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('one', () => {
    expect(one).toEqual(1)
   })
})
",
    "cursor": 123,
    "type": "insert",
  },
  {
    "char": " ",
    "content": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('one', () => {
    expect(one).toEqual(1)
    })
})
",
    "cursor": 124,
    "type": "insert",
  },
  {
    "char": " ",
    "content": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('one', () => {
    expect(one).toEqual(1)
     })
})
",
    "cursor": 125,
    "type": "insert",
  },
  {
    "char": " ",
    "content": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('one', () => {
    expect(one).toEqual(1)
      })
})
",
    "cursor": 126,
    "type": "insert",
  },
  {
    "char": "e",
    "content": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('one', () => {
    expect(one).toEqual(1)
    e  })
})
",
    "cursor": 127,
    "type": "insert",
  },
  {
    "char": "x",
    "content": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('one', () => {
    expect(one).toEqual(1)
    ex  })
})
",
    "cursor": 128,
    "type": "insert",
  },
  {
    "char": "p",
    "content": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('one', () => {
    expect(one).toEqual(1)
    exp  })
})
",
    "cursor": 129,
    "type": "insert",
  },
  {
    "char": "e",
    "content": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('one', () => {
    expect(one).toEqual(1)
    expe  })
})
",
    "cursor": 130,
    "type": "insert",
  },
  {
    "char": "c",
    "content": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('one', () => {
    expect(one).toEqual(1)
    expec  })
})
",
    "cursor": 131,
    "type": "insert",
  },
  {
    "char": "t",
    "content": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('one', () => {
    expect(one).toEqual(1)
    expect  })
})
",
    "cursor": 132,
    "type": "insert",
  },
  {
    "char": "(",
    "content": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('one', () => {
    expect(one).toEqual(1)
    expect(  })
})
",
    "cursor": 133,
    "type": "insert",
  },
  {
    "char": "t",
    "content": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('one', () => {
    expect(one).toEqual(1)
    expect(t  })
})
",
    "cursor": 134,
    "type": "insert",
  },
  {
    "char": "w",
    "content": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('one', () => {
    expect(one).toEqual(1)
    expect(tw  })
})
",
    "cursor": 135,
    "type": "insert",
  },
  {
    "char": "o",
    "content": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('one', () => {
    expect(one).toEqual(1)
    expect(two  })
})
",
    "cursor": 136,
    "type": "insert",
  },
  {
    "char": ")",
    "content": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('one', () => {
    expect(one).toEqual(1)
    expect(two)  })
})
",
    "cursor": 137,
    "type": "insert",
  },
  {
    "char": ".",
    "content": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('one', () => {
    expect(one).toEqual(1)
    expect(two).  })
})
",
    "cursor": 138,
    "type": "insert",
  },
  {
    "char": "t",
    "content": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('one', () => {
    expect(one).toEqual(1)
    expect(two).t  })
})
",
    "cursor": 139,
    "type": "insert",
  },
  {
    "char": "o",
    "content": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('one', () => {
    expect(one).toEqual(1)
    expect(two).to  })
})
",
    "cursor": 140,
    "type": "insert",
  },
  {
    "char": "E",
    "content": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('one', () => {
    expect(one).toEqual(1)
    expect(two).toE  })
})
",
    "cursor": 141,
    "type": "insert",
  },
  {
    "char": "q",
    "content": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('one', () => {
    expect(one).toEqual(1)
    expect(two).toEq  })
})
",
    "cursor": 142,
    "type": "insert",
  },
  {
    "char": "u",
    "content": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('one', () => {
    expect(one).toEqual(1)
    expect(two).toEqu  })
})
",
    "cursor": 143,
    "type": "insert",
  },
  {
    "char": "a",
    "content": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('one', () => {
    expect(one).toEqual(1)
    expect(two).toEqua  })
})
",
    "cursor": 144,
    "type": "insert",
  },
  {
    "char": "l",
    "content": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('one', () => {
    expect(one).toEqual(1)
    expect(two).toEqual  })
})
",
    "cursor": 145,
    "type": "insert",
  },
  {
    "char": "(",
    "content": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('one', () => {
    expect(one).toEqual(1)
    expect(two).toEqual(  })
})
",
    "cursor": 146,
    "type": "insert",
  },
  {
    "char": "2",
    "content": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('one', () => {
    expect(one).toEqual(1)
    expect(two).toEqual(2  })
})
",
    "cursor": 147,
    "type": "insert",
  },
  {
    "char": ")",
    "content": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('one', () => {
    expect(one).toEqual(1)
    expect(two).toEqual(2)  })
})
",
    "cursor": 148,
    "type": "insert",
  },
  {
    "char": "
",
    "content": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('one', () => {
    expect(one).toEqual(1)
    expect(two).toEqual(2)
  })
})
",
    "cursor": 149,
    "type": "insert",
  },
]
`;
